### ISSUE TITLE
Describing Hot Pluggable Memory (Bugzilla Bug 4062)

### ISSUE THREAD

**Issue by tianocore-issues** [tianocore/edk2#9201]
This issue was created automatically with [bugzilla2github](https://github.com/makubacki/bugzilla2github/tree/bz2gh_tianocore)
# Bugzilla Bug 4062
Date: 2022-09-15T02:13:02+00:00
From: Thanunathan Rangarajan &lt;<thanu.rangarajan>&gt;
To:   Samer El-Haj-Mahmoud &lt;<samer.el-haj-mahmoud>&gt;
CC:   @lgao4, Jose.Marinho, leo.duran, @mdroth, samer.el-haj-mahmoud, scolee, thanu.rangarajan

Last updated: 2023-04-30T10:02:59+00:00

**Comment by tianocore-issues** [comment]
## Comment 19568
Date: 2022-09-15 02:13:02 +0000
From: Thanunathan Rangarajan &lt;<thanu.rangarajan>&gt;

- Industry Specification: ACPI Specification
- Target OS: ---
- Bugzilla Assignee(s): Samer El-Haj-Mahmoud &lt;<samer.el-haj-mahmoud>&gt;

Created attachment 1024
Code-first SPM in SRAT

This ECR is related to a proposal to extend the ACPI SRAT to support hot-adding memory that is intended for specific-purpose usage, aka specific-purpose memory (SPM). The UEFI specification defines a memory attribute (EFI_MEMORY_SP) for describing SPM that is present and available at boot time. However, this attribute does not cover scenarios where the memory is unavailable/absent at boot time, and then hot-added physically or logically at runtime, and then used as SPM. Such dynamically added memory cannot be described in the EFI memory map and must be declared in the ACPI SRAT or HMAT tables instead. This ECR proposes that the SPM attribute be extended to the SRAT table.

An OS that is aware of this SRAT attribute is thus able to assign the hot-added memory to the correct memory pool - SPM or conventional memory (general-purpose).

Attachment: [BZ-1024-CodeFirst_SPM2_ver0.3.docx](https://github.com/tianocore/user-attachments/raw/refs/heads/main/tianocore/edk2/BZ-1024-CodeFirst_SPM2_ver0.3.docx)

**Comment by tianocore-issues** [comment]
## Comment 19651
Date: 2022-09-27 22:20:10 +0000
From: @lgao4

thanu.rangarajan@arm.com, please work on it.

**Comment by tianocore-issues** [comment]
## Comment 19652
Date: 2022-09-27 22:20:18 +0000
From: @lgao4

thanu.rangarajan@arm.com, please work on it.

**Comment by tianocore-issues** [comment]
## Comment 19683
Date: 2022-09-30 08:12:46 +0000
From: Samer El-Haj-Mahmoud &lt;<samer.el-haj-mahmoud>&gt;

ASWG mantis ECR: https://mantis.uefi.org/mantis/view.php?id=2354

**Comment by tianocore-issues** [comment]
## Comment 19772
Date: 2022-10-13 11:53:41 +0000
From: Thanunathan Rangarajan &lt;<thanu.rangarajan>&gt;

Created attachment 1037
Code-first - Hot Pluggable Memory Description

Attachment: [BZ-1037-CodeFirst_Hot Pluggable Memory_ver0.3.docx](https://github.com/tianocore/user-attachments/raw/refs/heads/main/tianocore/edk2/BZ-1037-CodeFirst_Hot%20Pluggable%20Memory_ver0.3.docx)

**Comment by tianocore-issues** [comment]
## Comment 19791
Date: 2022-10-19 10:51:43 +0000
From: Thanunathan Rangarajan &lt;<thanu.rangarajan>&gt;

Created attachment 1038
Code-first - Hot Pluggable Memory Description ECR v.0.65

Attachment: [BZ-1038-CodeFirst_SPM2_ver0.65.docx](https://github.com/tianocore/user-attachments/raw/refs/heads/main/tianocore/edk2/BZ-1038-CodeFirst_SPM2_ver0.65.docx)

**Comment by tianocore-issues** [comment]
## Comment 19792
Date: 2022-10-19 13:35:04 +0000
From: @mdroth

I agree with Proposal 1 in the latest attachment (0.65), but wouldn&apos;t it make sense to add a reference to SRAT? I&apos;d propose something akin to this (based on the latest 0.65 update):


9.x.1 Hot-plug Indication (Proposal 1)

If the memory device is created for the purpose of describing hot-pluggable memory, it must always carry the _STA, as well as either _EJ0 or _OST methods, or both. OS can use the presence of these methods as an indication that the memory range is hot-pluggable. In addition, there must be a matching memory affinity structure in the SRAT table that has the Hot-pluggable flag set. The expression for confirming hot-pluggable property is as follows:

Is Hot-pluggable = _STA &amp;&amp; (EJ0 || OST);



----

Also, in the SRAT section, Table 5.59 &quot;Flags - Memory Affinity Structure&quot; I would add a reference in the &quot;Hot Pluggable&quot; field description pointing back to 9.x.1 to further enforce the requirements associated with representing hot-pluggable memory.

**Comment by tianocore-issues** [comment]
## Comment 19794
Date: 2022-10-20 02:11:29 +0000
From: Thanunathan Rangarajan &lt;<thanu.rangarajan>&gt;

@Mike: thanks for the feedback. The SRAT is mentioned in section 9.x.2, titled Proximity Domains, but your suggested alternative is much better.

Also fixed some minor typos (e.g., _EJ0 instead of EJ0).

Updated ECR (v.0.7) attached for review.

**Comment by tianocore-issues** [comment]
## Comment 19795
Date: 2022-10-20 02:13:24 +0000
From: Thanunathan Rangarajan &lt;<thanu.rangarajan>&gt;

Created attachment 1039
Code-first - Hot Pluggable Memory Description ECR v.0.7

Attachment: [BZ-1039-CodeFirst_SPM2_ver0.7.docx](https://github.com/tianocore/user-attachments/raw/refs/heads/main/tianocore/edk2/BZ-1039-CodeFirst_SPM2_ver0.7.docx)

**Comment by tianocore-issues** [comment]
## Comment 19872
Date: 2022-10-28 20:20:56 +0000
From: Scott Lee &lt;<scolee>&gt;

&gt;&gt; •	Mandating the use of the ACPI memory device object for describing hot-pluggable memory, where the Extended Address Space resource descriptor is used to specify the memory attributes and memory type of the hot-pluggable address range. A reference ASL example is also provided. 

I have a question on above. Is intent for all types of hot-pluggable memory or limited to specific types of hot-pluggable memory? FYI ... the current definitions in CXL does not require the need for ACPI memory device object for hot-pluggable CXL memory. From the OS perspective, I do not want this new dependency for hot-pluggable CXL memory.

&gt;&gt; In the case of memory hot-add, the OS can then use the _ATT field information to understand how the memory must be used after it has been added. This enables hot-plug support for specific-purpose memory (SPM) and persistent memory. Since the _ATT field is optional, the OS must consider its absence to mean that the memory is by default cacheable memory with EFI attributes set to EFI_MEMORY_WB.
...
Hot-pluggable persistent memory ranges must not be described using this mechanism. They should instead be described using the NFIT table and related methods specific to persistent memory.

With the above text, I question how OS can rely on the definition given _ATT field is optional. The sentence &quot;This enables hot-plug support for specific-purpose memory (SPM) and persistent memory&quot; is incorrect as the last paragraph states that you cannot do this for persistent memory. My other concern with _ATT is that it places a dependency on availability of ASL. ASL isn&apos;t available all the time in the OS. This may result in additional work in the OS to retrieve this information and delay the processing of the hotplug event.

We have interest in being able to discover SPM from SRAT in all cases (including non-hotplug) as it simplifies the work needed to support SPM in our OS and hypervisor. We like the initial proposal of adding a new SPM bit in SRAT.

**Comment by tianocore-issues** [comment]
## Comment 20157
Date: 2022-11-30 17:04:15 +0000
From: Samer El-Haj-Mahmoud &lt;<samer.el-haj-mahmoud>&gt;

Created attachment 1056
Code-first - Hot Pluggable Memory Description ECR v.0.8

Attachment: [BZ-1056-Code-first - Hot Pluggable Memory Description ECR v.0.8.docx](https://github.com/tianocore/user-attachments/raw/refs/heads/main/tianocore/edk2/BZ-1056-Code-first%20-%20Hot%20Pluggable%20Memory%20Description%20ECR%20v.0.8.docx)

**Comment by tianocore-issues** [comment]
## Comment 20579
Date: 2023-01-10 13:46:00 +0000
From: Samer El-Haj-Mahmoud &lt;<samer.el-haj-mahmoud>&gt;

Following up on Scott&apos;s comment from 10/28/2022:

&quot;
We have interest in being able to discover SPM from SRAT in all cases (including non-hotplug) as it simplifies the work needed to support SPM in our OS and hypervisor. We like the initial proposal of adding a new SPM bit in SRAT.
&quot;

This is now being tracked as a separate code-first ECR from this one. The new ECR is:

Title: Exposing SPM in SRAT 
Bugzilla:  https://bugzilla.tianocore.org/show_bug.cgi?id=4138
ASWG Mantis: https://mantis.uefi.org/mantis/view.php?id=2361

**Comment by tianocore-issues** [comment]
## Comment 20618
Date: 2023-01-11 14:54:49 +0000
From: @mdroth

In the ACPI group we got a proposal on how to go about reflecting whether something is hot pluggable or not, and in the SRAT we do have one bit that claims that, but it&apos;s insufficient. One of the issues that was brought up is that when the system boots, the OS may have allocated memory in regions that are deemed hot pluggable (meaning it could be removed) and this would cause tragic situations for most OSes, especially if they&apos;d allocated any critical resources in those regions.

There&apos;s a desire to know early on what regions are deemed hotpluggable so that the OS in the future could get those hints and if possible avoid allocating critical resources in those regions.

In UEFI the current memory map always reflects the attributes associated with the memory that&apos;s been plugged in. So if you hypothetically have a region that&apos;s hotpluggable AND has memory populated it will show up in the memory map, but theoretically could go away. This could cause problems.

The only reasonable way to address this situation is an update in the memory map that&apos;s handed up to the OS, and we needed an attribute to decorate that information. This at least is backward compatible - doesn&apos;t cause any real chaos to anyone. However, in future operating systems and loaders, they could acknowledge such an attribute and judiciously do their allocations as needed without endangering themselves.

There&apos;s a proposal pending to add a new memory attribute to reflect whether a memory region reflects a hotpluggable capability.

Hope that makes sense...

**Comment by tianocore-issues** [comment]
## Comment 20619
Date: 2023-01-11 16:50:42 +0000
From: leo.duran

Actually,
I&apos;d argue that another &quot;reasonable&quot; and perhaps &quot;safer&quot; way is to simply exclude hot-pluggable ranges from the UEFI memory, and described them to the OS by other means (e.g., via memory device objects in the ACPI namespace).

I say &quot;safer&quot; because legacy OSes would not be aware of the proposed Hot-pluggable Attribute, hence could make undesirable decisions.

But perhaps I&apos;m missing a compelling reason to include hot-pluggable ranges in the UEFI memory map that are enabled at boot-time when ranges that are not-enabled at boot-time would still need to be declared by other means.

**Comment by tianocore-issues** [comment]
## Comment 20620
Date: 2023-01-11 18:21:55 +0000
From: @mdroth

There are some platforms that may have entire ranges capable of being hot plugged and in that case we cannot simply leave them out of the memory map. I believe we went over this scenario before - we have to comprehend that situation.

**Comment by tianocore-issues** [comment]
## Comment 20621
Date: 2023-01-11 22:11:07 +0000
From: @mdroth

Also, the scenario associated with the possibility of something being allocated on a hotpluggable range exists today already. We can&apos;t exactly fix that problem retroactively, but looking forward - it&apos;s possible. The presumption of course would be the OSes and/or loaders get updated with knowledge of the ACPI updates and new memory attribute bit.

**Comment by tianocore-issues** [comment]
## Comment 20623
Date: 2023-01-11 23:43:41 +0000
From: leo.duran

Nahh... (LOL)... Those platforms must have some &quot;fixed&quot; memory... I mean, it&apos;s impossible for all memory to &quot;go away&quot; during runtime, right?

So, it seems &quot;reasonable&quot; to declare some of it as &quot;fixed&quot; in the UEFI memory map, and the rest as &quot;hot-pluggable&quot; outside of the UEFI memory map.

**Comment by tianocore-issues** [comment]
## Comment 20624
Date: 2023-01-12 01:06:13 +0000
From: @mdroth

(In reply to LeoDuran@AMD from comment #16)
&gt; Nahh... (LOL)... Those platforms must have some &quot;fixed&quot; memory... I mean,
&gt; it&apos;s impossible for all memory to &quot;go away&quot; during runtime, right?
&gt; 
&gt; So, it seems &quot;reasonable&quot; to declare some of it as &quot;fixed&quot; in the UEFI
&gt; memory map, and the rest as &quot;hot-pluggable&quot; outside of the UEFI memory map.

You&apos;d think so - but I was told with some level of certainty that there would be support/examples of platforms that may have everything hotpluggable. I was as surprised as you when I heard that.

**Comment by tianocore-issues** [comment]
## Comment 20636
Date: 2023-01-12 11:33:45 +0000
From: leo.duran

Two points:
1) The OS needs fixed space for things like ISRs tables and data structures.
2) I presume that the proposed hot-pluggable attribute would serve as a hint to the OS to not use that memory for that fixed stuff.

So it seems like those 2 points are contradictory in the context of a system where all memory is tagged as hot-pluggable!

Basically, it&apos;s not clear to me the problem that the proposed hot-pluggable attribute is solving.

**Comment by tianocore-issues** [comment]
## Comment 21111
Date: 2023-02-23 11:30:51 +0000
From: leo.duran

Mike, et al,

My previous &quot;push-back&quot; comments on including hot-plug memory in the EFI map were based on the assumption that there would be an ACPI memory device object representing that memory. However, after thinking about this some more, there may be ways to add/remove memory during OS runtime that would *not* make use of memory device objects or platform firmware interactions.

Hence, in retrospect, I can see where having a UEFI hot-plug attribute could be a useful tag for memory that is present and usable at OS boot. (of course, memory not present at boot-time would *not* be included in the EFI map).

**Comment by tianocore-issues** [comment]
## Comment 21335
Date: 2023-04-06 08:02:50 +0000
From: Jose Marinho &lt;<Jose.Marinho>&gt;

Created attachment 1352
Code-first - Hot Pluggable Memory Description ECR v.0.9

Allow the hot-pluggable memory, described in the ECR, to also be present in the UEFI memory descriptors.

Attachment: [BZ-1352-Code-first - Hot Pluggable Memory Description ECR v.0.9.docx](https://github.com/tianocore/user-attachments/raw/refs/heads/main/tianocore/edk2/BZ-1352-Code-first%20-%20Hot%20Pluggable%20Memory%20Description%20ECR%20v.0.9.docx)

**Comment by tianocore-issues** [comment]
## Comment 21457
Date: 2023-04-30 10:02:09 +0000
From: Samer El-Haj-Mahmoud &lt;<samer.el-haj-mahmoud>&gt;

This ECR has been approved as-submitted in the UEFI Forum ASWG working group on Apr 27, 2023. The latest version of the ECR (v0.9) will be incorporated as new content in ACPI.next future. Since this is approved, implementations can proceed with the current approved definitions.

**Comment by tianocore-issues** [comment]
## Comment 21458
Date: 2023-04-30 10:02:59 +0000
From: Samer El-Haj-Mahmoud &lt;<samer.el-haj-mahmoud>&gt;

Approved for ACPI.next specification (from the &quot;code first&apos; public ECR definition in ver 0.9 of the attached document)
